### PR TITLE
Evaluate GitHub GPU Runners

### PR DESCRIPTION
**Pull request by glenn-jocher**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Major updates to GPU CI workflow for enhanced compatibility and efficiency.

### 📊 Key Changes
- **Expanded Testing Matrix**: Now includes `gpu-latest`, `github-windows11-gpu-t4-4-core`, and `github-linux-gpu-t4-4-core`, broadening the test environment. 🖥️
- **Custom Python Installation**: Moves from using the `actions/setup-python` action to a custom PowerShell script for installing Python, ensuring compatibility with specific Python versions. 🐍
- **7zip Installation**: A step to install 7zip on Windows environments has been added, enabling better handling of archives. 📦
- **Updated Python & Pip**: Introduction of Python 3.11 installation and pip upgrade commands to ensure the latest versions are used for testing. 🆕

### 🎯 Purpose & Impact
- **Broader Compatibility Testing**: The introduction of a more diverse matrix of operating systems and GPU configurations ensures that the software is tested in environments closer to what end-users might have. This should lead to fewer compatibility issues. 👥
- **Custom Python Install**: By controlling the Python installation process, the team can ensure that the precise Python versions needed for testing are used, leading to more reliable test outcomes. This might also impact the setup time slightly but guarantees version accuracy. 🎲
- **Efficiency and Reliability**: Updating to the latest Python and pip versions, along with the installation of necessary utilities like 7zip, aims at making the CI workflow more efficient and reliable, potentially reducing build times and increasing success rates. ⏱️🚀

These changes aim to improve the quality and reliability of Ultralytics software on various platforms by ensuring thorough testing and compatibility, leading to a smoother experience for end-users and developers alike.

### CODE REVIEW

**Comment by Burhan-Q** [comment]
@glenn-jocher looks like https://github.com/actions/setup-python/pull/827 _might_ be able to address the issue with the GitHub runner for Windows failing. I'm not familiar with how the GitHub actions are implemented, but is there anyway we can checkout the PR for this action to test with?

**Comment by glenn-jocher** [comment]
Hey there! 😊 

Thanks for bringing this to our attention and offering a potential solution! 👍 

To test with the changes from [actions/setup-python#827](https://github.com/actions/setup-python/pull/827), you can indeed modify the GitHub actions YAML file to use your branch directly. Here's how you can tweak it, based on your suggestion:

```yaml
- name: Set up Python
  uses: SMoraisAnsys/setup-python @SMoraisAnsys/no-pwshpath
```

Just replace the relevant part of the GitHub Action workflow with the code snippet above, and it should utilize the version from your PR. It looks promising, and we're hopeful it'll address the runner issue on Windows. 

Let's give it a shot, and huge thanks for your proactive approach! 🚀

**Comment by SMoraisAnsys** [comment]
@Burhan-Q Sorry I forwarded the wrong branch name ... The one I provided in my previous comment was used to reproduce the bug on github runners... The right branch to use is `fix/windows_extract_cpython_release` see https://github.com/actions/setup-python/pull/827
Sorry for the waste of time ...

TLDR: change workflow to `uses: SMoraisAnsys/setup-python@fix/windows_extract_cpython_release`

**Comment by Burhan-Q** [comment]
@SMoraisAnsys thank you! I tried it out and then when I was looking thru the original branch you shared, I realized that it was for replicating the issue 😆 so I tried playing around with updating PowerShell instead. This is the first time I've messed around with GitHub Actions, so it's been quite the learning experience!

**Comment by Burhan-Q** [comment]
Maybe I didn't setup the action correctly but seems that the error is the same on the `github-windows11-gpu-t4-4-core` runner

**Comment by SMoraisAnsys** [comment]
> Maybe I didn't setup the action correctly but seems that the error is the same on the `github-windows11-gpu-t4-4-core` runner

You seem to be good, could you try the call associated to ` [System.IO.Compression.ZipFile]::ExtractToDirectory('${escapedFile}', '${escapedDest}', $true) }` inside of your VM ? Btw, are you sure that there is a way to decompress the archive ? From my knowledge, setup-python action requires `7zip` see https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#windows

**Comment by Burhan-Q** [comment]
I'm confused. If 7zip is required for Windows runners for python, then why is it using the native `Expand-Archive` command? Also, it _looks_ like it's downloading a normal `.zip` file. In fact, I downloaded the linked file to my local Windows PC and it's a normal `.zip` file, then ran `Expand-Archive "python-3.11.9-win32-x64.zip"` so I'm confused as to why there's a problem.

**Comment by Burhan-Q** [comment]
Again, I'm quite naïve here, but it looks like (from the logs) that the script for unzipping the file is pointed at a directory `'C:\a\_temp\28ed0eb2-ba37-4036-a985-e36b60269f45'` and not an actual zip file.

**Comment by glenn-jocher** [comment]
@Burhan-Q hey I wouldn't worry about Windows too much here. The Windows issue seems to not be under our control and probably something we should email them directly about so they can fix on their end. We have Linux GPU runners already succeeding halfway up the PR so that's probably a simple place to start.

In any case this PR is a bit of an experiment to see how the GPU runners perform. They're much slower than our custom GPU runners so I'm not in love with GitHub charging us $0.07/min (linux) while the runner installs all the CUDA packages. 

Windows is higher at $0.14/min also.

Screenshot 2024-04-26 at 22 43 25

**Comment by glenn-jocher** [comment]
@Burhan-Q @SMoraisAnsys closing PR since we were able to run Linux GPU runners successfully. We'll keep them available for edge case testing and for consideration in the future. 

Thank you for your contributions!

**Comment by SMoraisDev** [comment]
@Burhan-Q (SMoraisAnsys here) When you download the file, it may not endup with a .zip naming but that doesn't make it a non zip file. However, from what I was able to understand in my attempt to fix the setup-python https://github.com/actions/setup-python/pull/819 the problem we are facing could be happening because the zip file was compressed using an unsupported compression method (e.g. through zip with Unix/Linux).
@glenn-jocher Feel free to ping me (SMoraisAnsys) if I can help you guys later on :)

**Comment by glenn-jocher** [comment]
Hey @SMoraisDev! Thanks for that insight. It's interesting how compression methods can impact compatibility across different systems. I'll experiment with the methods of decompression available in our environment to see if we can smoothly handle such cases. 🛠️  

We highly appreciate your willingness to assist! I might indeed ping you later as we debug and refine this setup. Meanwhile, if there are any pieces of code you think might help directly in this context, please do share! 👍

**Comment by SMoraisAnsys** [comment]
> Hey @SMoraisDev! Thanks for that insight. It's interesting how compression methods can impact compatibility across different systems. I'll experiment with the methods of decompression available in our environment to see if we can smoothly handle such cases. 🛠️
> 
> We highly appreciate your willingness to assist! I might indeed ping you later as we debug and refine this setup. Meanwhile, if there are any pieces of code you think might help directly in this context, please do share! 👍

I'll have a look when I got time then. In the mean time, could you up vote https://github.com/actions/setup-python/issues/819 to help getting the attention of the github actions developer? It's a pain that everyone has to perform manual work arounds to have a working CICD :/

**Comment by Burhan-Q** [comment]
I know we're not using the Windows GPU runners, but I wanted to post this here for future reference. Looks like there was a fix added https://github.com/actions/setup-python/issues/819#issuecomment-2275207223 and appears others are seeing it working.
